### ISSUE TITLE
How can I implement the YOLOv8 TFLite model with an output shape of [1, 9, 8400] in Android? 

### ISSUE THREAD

**Issue by chenchen-boop** [ultralytics/ultralytics#8218]
### Search before asking

- [ ] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I have find the same question on [https://github.com/ultralytics/ultralytics/issues/2950](url),but it can't slove my problem ,and it already closed ,so I asked the qusetion again.And there is the sourse seems could slove my problem [https://github.com/surendramaran/YOLOv8-TfLite-Object-Detector](url), but it is too hard to understand all to put into my project. 

I trained my custom dataset on yolov8, after finished training, I exported my model as .tfile. 

Model have 5 class.
input shape is [1,640,640,3], Type float32,
output shape is [1, 9, 8400], Type float32.

**In my classifyImage function, I want to get the highest class confidence and coordinate**, but I dont know how to do.

below is my core code: 
```
public void classifyImage(Bitmap image) {
        try {

            // Initialization code
            // Create an ImageProcessor with all ops required. For more ops, please
            // refer to the ImageProcessor Architecture section in this README.
            ImageProcessor imageProcessor =
                    new ImageProcessor.Builder()
                            .add(new ResizeOp(imageSize, imageSize, ResizeOp.ResizeMethod.BILINEAR))
                            .build();

            TensorImage tensorImage = new TensorImage(DataType.FLOAT32);


            // Preprocess the image
            tensorImage.load(image);
            tensorImage = imageProcessor.process(tensorImage);

            ModelV1 model = ModelV1.newInstance(getApplicationContext());
            // Runs model inference and gets result.
            ModelV1.Outputs outputs = model.process(tensorImage);
            Log.d("ModelV1.Outputs outputs;", "OK");

            TensorBuffer outputTensor = outputs.getOutputAsTensorBuffer();
            String[] classes = {"ball", "made", "person", "rim", "shoot"};

            // handle output (get the highest class confidence and coordinate)

  


 
          
            result.setText(classes[maxPos]);

            String s = "";
            for (int i = 0; i < classes.length; i++) {
                s += String.format("%s: %.1f%%\n", classes[i], confidences[i] * 100);
            }      
            confidence.setText(s);

            // Releases model resources if no longer used.
            model.close();

        } catch (Exception  e) {
            // TODO Handle the exception
            e.printStackTrace();
            Log.d(" classifyImage()", "Error: " + e.getMessage());
        }
    }
```

```
    public static Bitmap getResizedBitmap(Bitmap image, int newWidth, int newHeight) {
        return Bitmap.createScaledBitmap(image, newWidth, newHeight, true);
    }
```

Hope someone can help me. please.

**Comment by glenn-jocher** [comment]
@chenchen-boop hey there! 👋 It looks like you're trying to parse the output of your YOLOv8 TFLite model in Android. The output shape `[1, 9, 8400]` suggests that for each of the 8400 grid cells, the model predicts 9 values (which likely include class probabilities and bounding box coordinates).

Here's a simplified approach to process the output:

1. Reshape the output to `[8400, 9]` for easier handling.
2. For each row in the reshaped array:
   - Extract the bounding box coordinates.
   - Find the class index with the highest confidence score.
   - Retrieve the class name using the class index.

Here's a snippet that might help you get started:

```java
// Assuming 'outputTensor' is the model output with shape [1, 9, 8400]
float[][] outputArray = outputTensor.getFloatArray();

// Reshape the output to [8400, 9]
float[][] reshapedOutput = reshapeOutput(outputArray, 8400, 9);

// Process each prediction
for (float[] prediction : reshapedOutput) {
    // Extract bounding box coordinates (assuming they are the first 4 values)
    float x = prediction[0];
    float y = prediction[1];
    float width = prediction[2];
    float height = prediction[3];

    // Find the class with the highest confidence (assuming class confidences are the last 5 values)
    float maxConfidence = 0;
    int maxClassIndex = -1;
    for (int i = 4; i < prediction.length; i++) {
        if (prediction[i] > maxConfidence) {
            maxConfidence = prediction[i];
            maxClassIndex = i - 4; // Adjust index to get class index
        }
    }

    // Get class name using the index
    String className = classes[maxClassIndex];

    // Now you have the class name, confidence, and bounding box for this prediction
    // You can use this information as needed in your application
}

// Helper method to reshape the output
public float[][] reshapeOutput(float[] output, int rows, int cols) {
    float[][] reshaped = new float[rows][cols];
    for (int i = 0; i < rows; i++) {
        System.arraycopy(output, (i * cols), reshaped[i], 0, cols);
    }
    return reshaped;
}
```

Remember to adjust the indices based on how your model outputs the bounding box coordinates and class confidences. Also, ensure that the reshaping aligns with the output tensor's structure.

I hope this helps you move forward with your project! If you have further questions, feel free to ask. Good luck with your Android implementation! 🚀

**Comment by chenchen-boop** [comment]
First I need to thank for your reply, but I still have mistake. 

I tried my custom model and the result is that almost each grid cell get one of the classes super high confidence in any image   (above 90% ), after that I tried yolov8n.tflite to see whether is my model problem or not , and I got the same result (high confidence). 
I guess in my image preprocess have something got wrong, but I'm not sure.
Hope someone can help  me!!! 
```
 public void classifyImage(Bitmap image) {
        try {
       
            ImageProcessor imageProcessor =
                    new ImageProcessor.Builder()
                            .add(new ResizeOp(imageSize, imageSize, ResizeOp.ResizeMethod.BILINEAR))
                            .build();

            TensorImage tensorImage = new TensorImage(DataType.FLOAT32);

            // Analysis code for every frame
            // Preprocess the image
            tensorImage.load(image);
            tensorImage = imageProcessor.process(tensorImage);

            Yolov8n model= Yolov8n.newInstance(getApplicationContext());
            //ModelV1 model = ModelV1.newInstance(getApplicationContext());

            // Runs model inference and gets result.
            Yolov8n.Outputs outputs = model.process(tensorImage);
            
            TensorBuffer outputTensor = outputs.getOutputAsTensorBuffer();


          
            float[] outputArray = outputTensor.getFloatArray();

// Reshape the output to [8400, 9]
            //float[][] reshapedOutput = reshapeOutput(outputArray, 8400, 9);
            float[][] reshapedOutput = reshapeOutput(outputArray, 8400, 84);
//            float[] x = new float[8400];
//            float[] y = new float[8400];
//            float[] width = new float[8400];
//            float[] height = new float[8400];
//            float[] maxConfidence = new float[8400];
//            float[] className = new float[8400];

// Process each prediction
            String s = "";
            for (float[] prediction : reshapedOutput) {
                // Extract bounding box coordinates (assuming they are the first 4 values)
                float x = prediction[0];
                float y = prediction[1];
                float width = prediction[2];
                float height = prediction[3];

                // Find the class with the highest confidence (assuming class confidences are the last 5 values)
                float maxConfidence = 0;
                int maxClassIndex = -1;
                
                for (int i = 4; i < prediction.length; i++) {
                    if (prediction[i] > maxConfidence) {
                        maxConfidence = prediction[i];
                        maxClassIndex = i - 4; // Adjust index to get class index
                    }
                }
                if (maxClassIndex >-1 && maxConfidence > 0.7){

                    String className = classes[maxClassIndex];
                    s += String.format(Locale.US, "%s: %.1f%%\n", className, maxConfidence * 100);

                }

               Log.d("result",s);


            }

            confidence.setText(s);
          
            model.close();



        } catch (Exception  e) {
            
            e.printStackTrace();
            Log.d(" classifyImage()", "Error: " + e.getMessage());
        }
    }
```

**Comment by glenn-jocher** [comment]
@chenchen-boop implementing a YOLOv8 TFLite model in Android requires several steps, including preprocessing the input image, running the model inference, and post-processing the output to extract the class confidences and coordinates. The output shape `[1, 9, 8400]` suggests that the model returns a flattened list of detections, where each detection is represented by 9 values (e.g., class probabilities, bounding box coordinates, etc.), and there are 8400 such detections.

Here's a step-by-step guide to help you implement the model in your Android application:

1. **Preprocess the Input Image:**
   You've already handled this part in your `classifyImage` function, where you resize the image and convert it to a `TensorImage`.

2. **Run Model Inference:**
   You're also doing this correctly by calling `model.process(tensorImage)`.

3. **Post-process the Output:**
   This is the part where you need to interpret the output tensor. The output tensor shape `[1, 9, 8400]` likely contains the bounding box coordinates, class probabilities, and objectness scores for each detection. You'll need to reshape this tensor and apply non-maximum suppression (NMS) to filter out overlapping boxes and keep only the most confident detections.

Here's a pseudocode example of how you might post-process the output:

```java
// Assuming the output tensor shape is [1, 9, 8400]
float[][][] outputArray = outputTensor.getFloatArray();

// Define the number of classes and the number of attributes per detection
int numClasses = 5;
int attributesPerDetection = 9;

// Iterate over the detections and parse the output
List<Detection> detections = new ArrayList<>();
for (int i = 0; i < outputArray[0][0].length / attributesPerDetection; i++) {
    // Extract the attributes for the current detection
    float[] attributes = new float[attributesPerDetection];
    for (int j = 0; j < attributesPerDetection; j++) {
        attributes[j] = outputArray[0][j][i];
    }

    // Create a Detection object from the attributes (you'll need to define this class)
    Detection detection = new Detection(attributes);
    detections.add(detection);
}

// Apply NMS to filter the detections (you'll need to implement this method)
List<Detection> filteredDetections = applyNonMaximumSuppression(detections);

// Find the detection with the highest confidence
Detection highestConfidenceDetection = Collections.max(filteredDetections, Comparator.comparing(Detection::getConfidence));

// Get the class with the highest confidence
String highestConfidenceClass = classes[highestConfidenceDetection.getClassIndex()];
float highestConfidence = highestConfidenceDetection.getConfidence();

// Get the coordinates of the highest confidence detection
float[] coordinates = highestConfidenceDetection.getCoordinates();

// Update the UI with the highest confidence class and coordinates
result.setText(highestConfidenceClass);
confidence.setText(String.format("%s: %.1f%%", highestConfidenceClass, highestConfidence * 100));
```

In the pseudocode above, `Detection` is a custom class that you would need to define to hold the attributes of each detection, such as the bounding box coordinates, class index, and confidence score. The `applyNonMaximumSuppression` method is a standard algorithm used in object detection to filter out overlapping boxes based on their confidence scores and IoU (Intersection over Union) thresholds.

Please note that the exact indices for the class probabilities and bounding box coordinates within the `attributes` array will depend on how your model was trained and how the output tensor is structured. You'll need to adjust the parsing logic accordingly.

Lastly, the `applyNonMaximumSuppression` method is not trivial to implement, and you might want to look for existing libraries or code snippets that can perform NMS for you. Alternatively, you can perform NMS on the server side or in a more powerful environment if the mobile device's resources are limited.

Remember to test your implementation thoroughly to ensure that the detections are being parsed and filtered correctly.

**Comment by glenn-jocher** [comment]
@chenchen-boop hey there! It seems like you're encountering unusually high confidence scores across the board, which is indeed puzzling. 🤔 This could be due to several factors, but let's first ensure that the preprocessing is correct. 

Your preprocessing steps look fine, but it's crucial to normalize the pixel values to the range your model was trained on. If your model was trained with images normalized to [0, 1] or standardized, you should apply the same preprocessing here.

Also, the reshaping part seems to have a mismatch in dimensions. If your model output is `[1, 9, 8400]`, then reshaping to `[8400, 84]` is incorrect and should be `[8400, 9]` instead. This could be the reason for the high confidences, as the array might not be reshaped as intended, mixing up class confidences with other unrelated values.

Here's a quick check for normalization and reshaping:

```java
// Normalize pixel values if required (e.g., to [0, 1])
tensorImage = tensorImageProcessor.normalizeIfNeeded(tensorImage);

// Correct reshaping to [8400, 9]
float[][] reshapedOutput = reshapeOutput(outputArray, 8400, 9);

// Now process the reshaped output as before
```

Make sure to implement `normalizeIfNeeded` according to the normalization used during training. If these adjustments don't resolve the issue, it might be worth double-checking the model export process and ensuring that the TFLite model is correctly generated from the trained weights.

Keep me posted on your progress, and I'll be here to help you along the way!

**Comment by chenchen-boop** [comment]
@glenn-jocher Sorry to reply you so late.  
My training dataset is from robofIow ,I think I didnt't do anything like normalized or standardized .And my problem is still same -- high confidence . here is my code [https://github.com/chenchen-boop/AI_Ball_Detector.git](url).please help me

**Comment by glenn-jocher** [comment]
Hi @chenchen-boop! No worries about the timing. 😊 It's great that you're using a dataset from Roboflow; they usually handle normalization for you during the export process. 

Regarding the high confidence scores, it's possible that the issue might be with how the model's output is being interpreted. Since you haven't normalized the data during training, the preprocessing step should be fine as is.

I took a quick look at your repository, and I noticed that the reshaping part in your code still seems to be using `[8400, 84]`. If your model's output is `[1, 9, 8400]`, you'll want to reshape it to `[8400, 9]` to correctly interpret the results. Here's a small snippet to correct that:

```java
// Correct reshaping to [8400, 9]
float[][] reshapedOutput = reshapeOutput(outputArray, 8400, 9);
```

Make sure your `reshapeOutput` method is correctly implemented to handle this transformation. If you're still facing issues after this change, it might be helpful to revisit the model export process to ensure the TFLite model is correctly reflecting your trained weights.

Keep pushing forward, and let me know how it goes! 🚀

**Comment by chenchen-boop** [comment]
I thought I already changed reshape to［8400,9］in the link I provided.And the reshapeOutput function seems alright. Is there anything possible due to the high confidence?

**Comment by glenn-jocher** [comment]
@chenchen-boop hi there! 👋 If you've already adjusted the reshape to `[8400, 9]` and are still seeing high confidence scores, it might be worth checking the following:

1. **Model Export:** Ensure the model is exported correctly with the right parameters.
2. **Post-processing:** Verify that the output is being post-processed correctly, including applying any necessary sigmoid or softmax functions to the confidence scores if your model expects it.
3. **Thresholds:** Double-check that you're applying the correct confidence and non-maximum suppression (NMS) thresholds when interpreting the results.

If the issue persists, it could be helpful to inspect the raw model outputs before any reshaping or processing to see if they're as expected. Keep up the great work, and I'm sure you'll crack it! 🛠️

**Comment by chenchen-boop** [comment]
@glenn-jocher 
1. I noticed the image I capture from camera is too small (180x240),so I thought It is the major reason cause the weird confidence(extreme high or low). Should I use CameraX or Camera2 to get the higher capture resolution?
2. . I want to test my input whether or not need to be normalize , so at first I didn't use normalize getting extreme high confidence ,and I use normalize getting extreme low confidence.I don't know which did I trust. (Maybe I need to finish first problem and test it again )   
![螢幕擷取畫面 2024-02-28 204445](https://github.com/ultralytics/ultralytics/assets/85823375/50fa9355-e739-4259-8157-9606f8de7013)
![螢幕擷取畫面 2024-02-28 204827](https://github.com/ultralytics/ultralytics/assets/85823375/2a6f7b45-904e-4511-a3ca-76e9e34d2787)

**Comment by glenn-jocher** [comment]
@chenchen-boop hey there! 😊

1. Absolutely, using CameraX or Camera2 for higher resolution images can significantly improve model performance. Higher resolution images provide more detail, which can help reduce those extreme confidence scores you're seeing.

2. Regarding normalization, it's crucial to match the preprocessing steps used during training. If your model was trained with normalized inputs, you should normalize your test inputs similarly. The extreme confidence scores might be due to the mismatch in preprocessing. Here's a quick example of normalization:

```java
// Normalize pixel values to [0,1]
tensorImage.load(image);
float mean = 0.0f;
float std = 255.0f;
tensorImage = TensorImageUtils.normalize(tensorImage, mean, std);
```

Try addressing the resolution first, and then experiment with normalization to see which approach aligns best with your training setup. Keep up the great work! 🚀
